### DISCUSSION TITLE
Java vs OpenJDK usage

### DISCUSSION THREAD

**Discussion by rjkroll** [nextgenhealthcare/connect#5631] [Q&A] [answered]
Hi,

We are upgrading our Mirth server to resolve security vulnerabilities, which means upgrading Mirth to 4.1.2, as well as Postgres and Java.

They want us to upgrade to the 'latest version of Java', but I know there was some licensing changes recently.

If we were to go with the 'latest version of Java', do we still have to do the configuration as before, when using Java 9+ in the options files?

If we go with OpenJDK, is there any configuring needed there since MIrth 4.1.2 was released?  How do we tell it to use OpenJDK, through the Launcher?

Thanks in advance.

**Comment by pacmano1** [comment, marked as answer]
You still need java 9+ options.   OpenJDK is fine.   Dist is your call, e.g. Azul, Corretto, etc.    The launcher has an option in the gui to select what version (file path) you want to use for the mirth admin tool when launched.

**Reply by gbhoffman** [reply]
What controls which Java file path is used by mcservice.exe?

**Reply by ab-20-m** [reply]
Combination of registry entries and environment variables. The actual search path exe4j is less than well defined but usually:
1. ./jre  - (Undocumented but seems to be first search entry?)
2. ./.install4j/jre.bundle - (Undocumented but seems to be second search entry?)
3. Windows Registry for well known JVMs (depending on the version of exe4j + JVM used keys)
4. %JAVA_HOME%
5. %JDK_HOME%
6. %EXE4J_JVM_HOME% - Seems to be implementation detail and might take precedence over the above entries? Try at your own risk.

From the exe4j documentation:

> It is possible to generate a log file that contains information about the JRE search sequence and
> any potential problems. In order to switch on logging, start the executable with the
> /create-i4j-log argument. The launcher will notify the user where the log is created and will
> offer to open an explorer window with the log file selected. After the message box, the launcher
> will continue to start up. If it is not possible to pass arguments, define the environment
> variable EXE4J_LOG=yes and look for the newest text file whose name starts withi4j_nlog_ in
> the Windows %TEMP% directory.
> If the entire search sequence fails, exe4j will try the location defined by the environment variable
> EXE4J_JAVA_HOME. If that fails too, an error message will be displayed asking the user to define
> this variable. To supply a custom variable, define an appropriate environment variable search
> sequence entry and customize the corresponding error message in the "Messages" step of the
> wizard.

**Reply by rjkroll** [reply]
For anyone questioning what the steps are to move Mirth from using Java to OpenJDK, this is what we had to do:

-  Unzip latest OpenJDK to a local drive
-  Set/add the following Enviroemental Variables and point to the OpenJDK folder
-  Add JAVA_HOME, or point it to OpenJDK folder, EX4J_JAVA_HOME (same thing) and add the OpenJDK \bin folder to the PATH variable
- Uninstall previous versions of Java
-  Copy the code from the mcservice, or mcserver, .options file in the doc folder to the root folder version of the same file, this will allow Mirth to use Java 9+ (if you haven't already done this in the past)
-  When launching from the Mirth Admin Launcher, select 'Default' as the Java selection, which should now be OpenJDK

After you launch, in the dashboard or in the About, you should see it show that Mirth is now using OpenJDK 19.2 (or whatever version you downloaded)

**Comment by rjkroll** [comment]
Hi,

Had a follow-up question on using Mirth with OpenJDK.   I believe we did everything correctly, including setting the Enviromental Variables, as well as pointing to the OpenJDK path when launching Mirth through the Launcher.

In the Launcher, though, do we have to select 'Custom' and point to the OpenJDK path each time?   Is there an easier way to get it to know to use OpenJDK instead of regular Java, which is also still installed for now?   I thought I saw a screenshot somewhere of MIrth allowing you to select OpenJDK, or Java from a dropdown before launching.

When I login to Mirth and go to the About section, it does properly list Java 19.2, which is the latest version of OpenJDK we just installed.

**Reply by pacmano1** [reply]
I rarely change the version of java the mirth admin launcher bundles.   That one is fine from a license perspective - I think Nextgen addressed that.  But you can save the connection with the alternative jre specified.
